### DISCUSSION TITLE
(RuntimeError) The tesla adapter has not been configured.

### DISCUSSION THREAD

**Discussion by Rim-777** [visciang/telegram#172] [General]
Hello

I tried  to run the following example within the Phoenix project and got the following error:
Example:
```defmodule CountChatBot do
  @moduledoc false

  require Logger

  use Telegram.ChatBot

  @session_ttl 60 * 1_000

  @impl Telegram.ChatBot
  def init(_chat) do
    count_state = 0
    {:ok, count_state, @session_ttl}
  end

  @impl Telegram.ChatBot
  def handle_update(%{"message" => %{"chat" => %{"id" => chat_id}}}, token, count_state) do
    count_state = count_state + 1

    Telegram.Api.request(token, "sendMessage",
      chat_id: chat_id,
      text: "Hey! You sent me #{count_state} messages"
    )

    {:ok, count_state, @session_ttl}
  end

  def handle_update(update, _token, count_state) do
    Logger.info("Unknown update received: #{inspect(update)}")

    {:ok, count_state, @session_ttl}
  end
end

token = "6999904422:AAFueF9-djzcUgbt8XuV52qmeR9k-A5PIkk"

if token == nil do
  IO.puts(:stderr, "Please provide a BOT_TOKEN environment variable")
  System.halt(1)
end

{:ok, _} =
  Supervisor.start_link(
    [{Telegram.Poller, bots: [{CountChatBot, token: token, max_bot_concurrency: 1_000}]}],
    strategy: :one_for_one
  )

Process.sleep(:infinity)
```

```== Compilation error in file <file..>
** (exit) shutdown: failed to start child: Telegram.Poller
    ** (EXIT) an exception was raised:
        ** (RuntimeError) The tesla adapter has not been configured. This will defaults to the built-in erlang :httpc module.

Please configure a production ready client, for instance:

config :tesla, adapter: {Tesla.Adapter.Hackney, [recv_timeout: 40_000]}

            (telegram 1.2.1) lib/poller.ex:55: Telegram.Poller.assert_tesla_adapter_config/0
            (telegram 1.2.1) lib/poller.ex:32: Telegram.Poller.start_link/1
            (stdlib 5.2) supervisor.erl:420: :supervisor.do_start_child_i/3
            (stdlib 5.2) supervisor.erl:406: :supervisor.do_start_child/2
            (stdlib 5.2) supervisor.erl:390: anonymous fn/3 in :supervisor.start_children/2
            (stdlib 5.2) supervisor.erl:1258: :supervisor.children_map/4
            (stdlib 5.2) supervisor.erl:350: :supervisor.init_children/2
            (stdlib 5.2) gen_server.erl:980: :gen_server.init_it/2```


Could you please help me find a solution?
I will appreciate any help you can provide

**Comment by visciang** [comment]
@Rim-777 can you elaborate on "I tried to run the following example within the Phoenix project" ?
 
 If I got it right the script you pasted was taken from `telegram/example/example_chatbot.exs`.
 These scripts are simple self-contained (toy) examples that can be run just as elixir scripts as described in the [quick start](https://github.com/visciang/telegram?tab=readme-ov-file#quick-start) doc:
 
 `BOT_TOKEN="..." example/example_chatbot.exs`

**Reply by Rim-777** [reply]
We have an Elixir Phoenix API that interacts with external services. Currently, we need to integrate several Telegram bots built with Python. We want to avoid redundant HTTP calls, so we prefer not to have these external Telegram bots communicate with our API, but make them a part of the API so that we can avoid the redundant HTTP call since it significanly increases the latency of responses. 

To ensure this library is fully functional, I decided to test it within our Phoenix API. As a first step, I, indeed, used your example and tried to test it inside our Phoenix API. 

Here’s how I installed it:
I added this to `mix.exs deps/0`. 
`{:telegram, github: "visciang/telegram"}`

However, when I tried to run the API, I encountered the mentioned error. I also tried including `Telegram.Poller` in the supervisor's children list in Application.start/2, but I got the same result.

What did I do wrong? Could you provide an example of how to use the library inside a Phoenix application?

**Comment by visciang** [comment]
OK, then as you can see in the log error:

```
** (exit) shutdown: failed to start child: Telegram.Poller
    ** (EXIT) an exception was raised:
        ** (RuntimeError) The tesla adapter has not been configured. This will defaults to the built-in erlang :httpc module.

Please configure a production ready client, for instance:

config :tesla, adapter: {Tesla.Adapter.Hackney, [recv_timeout: 40_000]}
```

Make sure to configure the `config :tesla, adapter: ...` of your choice and the dependency installed.

This is described in the docs under https://github.com/visciang/telegram?tab=readme-ov-file#telegram-client-config
